From e73f0bffb26c59b5b3ff024b4194a3c6fe34a8dd Mon Sep 17 00:00:00 2001
From: Zhu Donghong <zhudonghong@wxiat.com>
Date: Tue, 21 Jun 2022 14:47:33 +0800
Subject: [PATCH] sw64: deliver a hot reset to Root Complex with plugin JMicron
 585 card
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: e73f0bffb26c59b5b3ff024b4194a3c6fe34a8dd
Modified-by-SEL: No


Sunway inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I5GFPA

--------------------------------

On SW64 platform, JMicron 585 SATA card directly connected to Root
Complex may raise DMA failure when reboot, so we force a hot reset
to Root Complex to fix can not access JMicron 585 SATA card.

Signed-off-by: Zhu Donghong <zhudonghong@wxiat.com>

Signed-off-by: Gu Zitao <guzitao@wxiat.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/sw_64/platform/platform_xuelang.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/arch/sw_64/platform/platform_xuelang.c b/arch/sw_64/platform/platform_xuelang.c
index 63a4b163e43e..ae8179b53b4c 100644
--- a/arch/sw_64/platform/platform_xuelang.c
+++ b/arch/sw_64/platform/platform_xuelang.c
@@ -26,9 +26,25 @@ extern void cpld_write(uint8_t slave_addr, uint8_t reg, uint8_t data);
 
 static void xuelang_kill_arch(int mode)
 {
+	struct pci_dev *pdev;
+	struct pci_controller *hose;
+	int val;
+
 	if (is_in_host()) {
 		switch (mode) {
 		case LINUX_REBOOT_CMD_RESTART:
+			pdev = pci_get_device(PCI_VENDOR_ID_JMICRON,
+					      0x0585, NULL);
+			if (pdev) {
+				hose = (struct pci_controller *)pdev->sysdata;
+				val = read_rc_conf(hose->node, hose->index,
+						   RC_PORT_LINK_CTL);
+				write_rc_conf(hose->node, hose->index,
+					      RC_PORT_LINK_CTL, val | 0x8);
+				write_rc_conf(hose->node, hose->index,
+					      RC_PORT_LINK_CTL, val);
+			}
+
 			cpld_write(0x64, 0x00, 0xc3);
 			mb();
 			break;
-- 
2.33.0

